### PR TITLE
fix: iteration

### PR DESCRIPTION
**Pull request by QuantumExplorer**
There were some big issues in iterators, this should fix them.

### CODE REVIEW

File: merk/src/proofs/query/mod.rs [hunk 1]
```diff
@@ -437,80 +437,100 @@ impl QueryItem {
...
-                let valid = iter.valid() && iter.key().is_some() && work;
-                let next_valid = iter.key()
-                    != Some(if left_to_right {
-                        range_inclusive.end()
-                    } else {
-                        range_inclusive.start()
-                    });
-                (valid, next_valid)
+                let basic_valid = iter.valid() && iter.key().is_some();
+                let valid = basic_valid && if left_to_right {
```
Comment (PastaPastaPasta) [inline merk/src/proofs/query/mod.rs]:
```
basic_valid && 
(
    (left_to_right && iter.key() <= Some(range_inclusive.end())) ||
     iter.key() >= Some(range_inclusive.start())
)
```

File: merk/src/proofs/query/mod.rs [hunk 2]
```diff
@@ -437,80 +437,100 @@ impl QueryItem {
...
-                    && work;
-                let next_valid = left_to_right || iter.key() != Some(start);
-                (valid, next_valid)
+                    && iter.key().is_some();
+                let valid = basic_valid && if left_to_right {
+                    true
+                } else {
+                    iter.key() >= Some(start)
+                };
+                valid
```
Comment (PastaPastaPasta) [inline merk/src/proofs/query/mod.rs:516]: simply
```
(basic_valid && (left_to_right || iter.key() >= Some(start))
```

File: merk/src/proofs/query/mod.rs [hunk 3]
```diff
@@ -437,80 +437,100 @@ impl QueryItem {
...
-                    && iter.key().is_some()
-                    && (!left_to_right || iter.key() != Some(end));
-                (valid, true)
+                    && iter.key().is_some();
+                let valid = basic_valid && if left_to_right {
+                    iter.key() < Some(end)
+                } else {
+                    true
+                };
+                valid
```
Comment (PastaPastaPasta) [inline merk/src/proofs/query/mod.rs:527]: same

File: merk/src/proofs/query/mod.rs [hunk 4]
```diff
@@ -437,80 +437,100 @@ impl QueryItem {
...
-                (valid, next_valid)
+                let basic_valid  = (limit == None || limit.unwrap() > 0)
+                    && iter.valid()
+                    && iter.key().is_some();
+                let valid = basic_valid && if left_to_right {
+                    iter.key() <= Some(end)
+                } else {
+                    true
+                };
+                valid
```
Comment (PastaPastaPasta) [inline merk/src/proofs/query/mod.rs:538]: same

File: merk/src/proofs/query/mod.rs [hunk 5]
```diff
@@ -437,80 +437,100 @@ impl QueryItem {
...
-                    && iter.key().is_some()
-                    && (left_to_right || iter.key() != Some(start));
-                (valid, true)
+                    && iter.key().is_some();
+                let valid = basic_valid && if left_to_right {
+                    true
+                } else {
+                    iter.key() > Some(start)
+                };
+                valid
```
Comment (PastaPastaPasta) [inline merk/src/proofs/query/mod.rs:549]: same

File: merk/src/proofs/query/mod.rs [hunk 6]
```diff
@@ -437,80 +437,100 @@ impl QueryItem {
...
-                    && (left_to_right || iter.key() != Some(start))
-                    && !(left_to_right && iter.key() == Some(end));
-                (valid, true)
+                    && iter.key().is_some();
+                let valid = basic_valid && if left_to_right {
+                    iter.key() < Some(end)
+                } else {
+                    iter.key() > Some(start)
+                };
+                valid
```
Comment (PastaPastaPasta) [inline merk/src/proofs/query/mod.rs:560]: same

File: merk/src/proofs/query/mod.rs [hunk 7]
```diff
@@ -437,80 +437,100 @@ impl QueryItem {
...
-                let next_valid = (left_to_right || iter.key() != Some(range_inclusive.start()))
-                    && !(left_to_right && iter.key() == Some(range_inclusive.end()));
-                (valid, next_valid)
+                    && iter.key().is_some();
+                let valid = basic_valid && if left_to_right {
+                    iter.key() <= Some(range_inclusive.end())
+                } else {
+                    iter.key() > Some(range_inclusive.start())
+                };
+                valid
```
Comment (PastaPastaPasta) [inline merk/src/proofs/query/mod.rs:587]: same

File: merk/src/proofs/query/mod.rs [hunk 8]
```diff
@@ -437,80 +437,100 @@ impl QueryItem {
...
-                // the first element
-                let next_valid = left_to_right || iter.key() != Some(start);
-                (valid, next_valid)
+                    && iter.key().is_some();
+                let valid = basic_valid && if left_to_right {
+                    iter.key() < Some(end)
+                } else {
+                    iter.key() >= Some(start)
+                };
+                valid
```
Comment (PastaPastaPasta) [inline merk/src/proofs/query/mod.rs:490]: simplify

File: merk/src/proofs/query/mod.rs [hunk 9]
```diff
@@ -1,10 +1,6 @@
 mod map;
 
-use std::{
-    cmp::{max, min, Ordering},
-    collections::BTreeSet,
-    ops::{Range, RangeFrom, RangeFull, RangeInclusive, RangeTo, RangeToInclusive},
-};
+use std::{cmp::{max, min, Ordering}, cmp, collections::BTreeSet, ops::{Range, RangeFrom, RangeFull, RangeInclusive, RangeTo, RangeToInclusive}};
```
Comment (PastaPastaPasta) [inline merk/src/proofs/query/mod.rs:3]: nit: imo this is worse 🤷

**Review by PastaPastaPasta — COMMENTED** [review]
LGTM, a couple things for future potential refactoring

File: merk/src/proofs/query/mod.rs (hunk 9 above)
Comment (fominok) [reply merk/src/proofs/query/mod.rs:3]: hmm a single `cargo fmt` will bring it back to normal

**Review by fominok — APPROVED** [review]

**Pull request merged**